### ISSUE TITLE
[Suggestion] Put built in vars in another category

### ISSUE THREAD

**Issue by thegamerx1** [zero-plusplus/vscode-autohotkey-debug#88]
Move built in vars to another category away from `Global` because they are fill 60% of the category with rarely used data.

![image](https://user-images.githubusercontent.com/34986564/105889290-4ea24c80-600e-11eb-9221-1126c0563599.png)

**Comment by zero-plusplus** [comment]
I am using the DeepL translator.

Thanks for the suggestion.
I checked to see if this could be implemented. As a result, the implementation was possible, but there were some problems.
![#88-1](https://user-images.githubusercontent.com/24719310/105943950-ae980200-60a5-11eb-99b3-b84040767318.png)

I'm currently working on another extension that supports AutoHotkey, so I don't have time for this.
Please be patient.

**Comment by zero-plusplus** [comment]
I want to show not only built-in variables, but also class categories, static variable categories and etc.
To solve these flexibly, I decided to implement user-defined categories.
This will allow the user to freely define and select the categories to be displayed.

**Comment by SAbboushi** [comment]
I think I like this.  

These are some thoughts that jumped into my head.  I have not fully considered them so they may not be good ideas, but I am providing for your consideration:
1) Category for Super-Globals with include/exclude based upon variable name
2) Sometimes I'm debugging someone's code and they've polluted the global / super-global scope. Maybe ability to categorize variables per #Include file? e.g. If #Include 

![image](https://user-images.githubusercontent.com/1964198/106011140-ca88ad00-607f-11eb-9186-9a13620eec67.png)

I'd prefer they are not listed here:
![image](https://user-images.githubusercontent.com/1964198/106011162-cf4d6100-607f-11eb-82c8-f133b282eb35.png)

3) How about ability to nest variables e.g. using example above: "EVENT" node expands to show "EVENT_CONSOLE" node expands to show the variables "EVENT_CONSOLE_..."

**Comment by zero-plusplus** [comment]
> 1. Category for Super-Globals with include/exclude based upon variable name

This will be difficult because there is no way to check whether a variable is a super-global or not. It could be implemented in the future as it is probably guessable using a script parser.

> 2. Sometimes I'm debugging someone's code and they've polluted the global / super-global scope. Maybe ability to categorize variables per #Include file? e.g. If #Include

This is difficult to implement because the debugger does not provide information on which file the global variable was defined in. However, similar to the above, a script parser would be a solution.

> 3. How about ability to nest variables e.g. using example above: "EVENT" node expands to show "EVENT_CONSOLE" node expands to show the variables "EVENT_CONSOLE_..."

I plan to provide regular expression matching, so it is possible to list only variables with a specific name.

**Comment by SAbboushi** [comment]
Thanks for sharing your thoughts.  

>>  regular expression matching
Nice!

**Comment by zero-plusplus** [comment]
I'm working on this issue and I'm having some trouble with it.

There are two candidates for matching variable names: regular expressions and wildcards.
Initially, I was thinking of using regular expressions, but I wonder if there are cases where the matching is that complicated.

There was a suggestion above to categorize variables that are all uppercase, but this can be handled by adding an `upper` attribute that determines if the variable is uppercase or not.

Please let me know your opinion.

**Comment by zero-plusplus** [comment]
Thanks for your opinion.

**Comment by zero-plusplus** [comment]
I am asking for your opinion again.

Some things cannot be categorized in different versions of AutoHotkey.
For example, functions only show up in data inspection in v2, so if a category is created for built-in functions only, nothing will show up in v1.

In this case, unnecessary categories will be displayed. I want to solve this issue.

So I came up with two ideas.
1. a way to switch depending on the version, such as `runtime_v2`.
2. If no variables are displayed in a category, hide them. (It might be better to make this an option so that users are not confused.)

Method 1 requires duplicate descriptions of categories that you do not want to hide, so we think method 2 is superior.
Or there may be another, better way.

If you have an opinion, please comment.

---

Here are my notes for implementing method 2.
* `hidden` - can be boolean or `"auto"`. You can show/hide it temporarily. If `"auto"` is specified, the category variable will be hidden when it is 0.

**Comment by zero-plusplus** [comment]
The following is an excerpt of the variableCategories description from the wiki, which is not publicly available. (It contains links that only work on wiki. e.g. `[[data inspection]]`)
<table>
<tr>
    <th>Name</th>
    <th>Type</th>
    <th>Description</th>
</tr>
<tr>
    <td rowspan="2">variableCategories</td>
    <td>string</td>
    <td>
        Specifies <code>"recommend"</code>. It divides [[data inspection]] groups into three categories: <code>Local</code>, <code>Global</code>, and <code>Built-in Global</code>. Also, the 0-9 deprecated variables that indicate arguments will not be displayed.
    </td>
</tr>
<tr>
    <td>array</td>
    <td>
        <strong>The <code>[Category]</code> and <code>[Matcher]</code> described below is a preview version and its use may change in the future.</strong><br /><br />
        Specify an array of <code>[Category]</code> objects with the following attributes.<br /><br />
    This is a very complex attribute, so practical examples are provided [[here|variableCategories-examples]].
    </td>
</tr>
<tr>
    <td><code>[Category]</code></td>
    <td>string</td>
    <td>An object that represents a group of variables. It has the following attributes.</td>
</tr>
<tr>
    <td><code>[Category].label</code></td>
    <td>string</td>
    <td>The title displayed in the group.</td>
</tr>
<tr>
    <td rowspan="2"><code>[Category].source</code></td>
    <td>string</td>
    <td>
        Specifies <code>"Local"</code>, <code>"Global"</code>.<br /><br />
    If you are using <a href="https://hotkeyit.github.io/v2/">AutoHotkey_H</a>, additionally you can specify <code>"Static"</code>.<br />
    If you want to include all of them, specify <code>"*"</code>.
    </td>
</tr>
<tr>
    <td>array</td>
    <td>Array of strings that can be specified above.</td>
</tr>
<tr>
    <td><code>[Category].matchers</code></td>
    <td>array</td>
    <td>An array of <code>[Matcher]</code> objects with the following attributes.</td>
</tr>
<tr>
    <td><code>[Matcher]</code></td>
    <td>object</td>
    <td>An object representing a match condition. If multiple conditions are specified, all of them will be matched or not. It has the following attributes.</td>
</tr>
<tr>
    <td><code>[Matcher].method</code></td>
    <td>string</td>
    <td>If you want to filter variable that match the conditions, specify "exclude". Otherwise, omit this attribute or specify "include".</td>
</tr>
<tr>
    <td><code>[Matcher].pattern</code></td>
    <td>string</td>
    <td><a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions">JavaScript RegExp patterns</a> for variable name.</td>
</tr>
<tr>
    <td><code>[Matcher].ignorecase</code></td>
    <td>boolean</td>
    <td>If <code>true</code>, <code>[Matcher].pattern</code> will ignore case.</td>
</tr>
<tr>
    <td><code>[Matcher].type</code></td>
    <td>string</td>
    <td>The type name recognized by the debugger. See [[here|Conditional-breakpoint#Basic-type-check]] for details.</td>
</tr>
<tr>
    <td><code>[Matcher].className</code></td>
    <td>string</td>
    <td>The class name recognized by the debugger. For more information, see CLASSNAME [[here|Conditional-breakpoint#Additional-type-check]].</td>
</tr>
<tr>
    <td><code>[Matcher].builtin</code></td>
    <td>boolean</td>
    <td>Whether it is a built-in variable or not.</td>
</tr>
<tr>
    <td><code>[Matcher].static</code></td>
    <td>boolean</td>
    <td>Whether it is a static variable or not.</td>
</tr>
</table>


The following are the variableCategories that will be set when `recommend` is specified.
```
"variableCategories": [
      {
        label: 'Local',
        source: [ 'Local', 'Static' ],
      },
      {
        label: 'Global',
        source: 'Global',
        matchers: [
          {
            method: 'exclude',
            builtin: true,
          },
        ],
      },
      {
        label: 'Built-in Global',
        source: 'Global',
        matchers: [
          { builtin: true },
          { method: 'exclude', pattern: '^\\d+$' },
        ],
      },
    ];
```

**Comment by SAbboushi** [comment]
> For example, functions only show up in data inspection in v2

I am feeling stupid because I do not know what you mean by this.  If you can give further details or screenshot, maybe I can respond to your request for opinions.

I had some difficulty understanding the excerpt, but your design seems to be flexibile and comprehensive to define variableCategories and I think it will allow me to create additional variableCategories to solve the pollution I refer to in my Jan 27 post above.  But I think your design does not allow for the nesting I described, correct?

Also, I suggest you rename "borrowMatchers" to "useMatcher".  I like your design to reuse already-defined matchers.

**Comment by zero-plusplus** [comment]
> > For example, functions only show up in data inspection in v2

> I am feeling stupid because I do not know what you mean by this. If you can give further details or screenshot, maybe I can respond to your request for opinions.

For example, built-in functions such as ObjGetBase and SubStr are not shown in data inspection in v1, but only in v2.

> But I think your design does not allow for the nesting I described, correct?

I was wrong about the nesting.

In vscode, the root of the data inspection (Local, Global, etc.) is distinguished as a scope, and its children as variables.

Since variableCategories is an attribute to define about the scope, it is not something that can be set about its children (i.e. variables).

In other words, you can create an Event scope and group variables starting with Event in it, but you cannot put an Event scope inside another scope.

It is technically possible, but I think it would be unacceptably complicated.

> Also, I suggest you rename `borrowMatchers` to `useMatcher`. I like your design to reuse already-defined matchers.

I'm glad to hear that my idea was good.

I had made the same mistake, but it is the `[Category]` object that has the named (label), not the `[Matcher]` object.
So I named it borrowMatchers because it borrows the matchers from `[Category]`.
The reason for the plural form is that it borrows all Matchers.

**Comment by zero-plusplus** [comment]
I noticed that `borrowMatchers` is a complicated and not so good attribute for both implementers and users.

For this reason, I propose `[Category].without`, which has more limited functionality.

It specifies the category name as an array. It removes variables in the specified category from the category.

**Comment by zero-plusplus** [comment]
I've been thinking about `borrowMatchers` and `without` for the past few days, and I'm beginning to think that neither is a very good solution.

These require a new attribute to be added each time a new category is added.
This may reduce the amount of description by a few lines, but it does not reduce the fundamental effort.

To solve this issue, I need a different approach. 

For example, the `noduplicate` attribute that prohibits duplication with other categories that have the same source.

**Comment by SAbboushi** [comment]
I wish you good luck with your continued efforts.  It is amazing to me how many things need to be considered in order to create an excellent debugger.  Thanks for your efforts.

**Comment by zero-plusplus** [comment]
Implemented `noduplicate`.
The limitation is that only one can be set for the same source.

This can be used, for example, to categorize certain variables (built-in variables, class objects, etc.) from the Global source, and then create a category without them.

**Comment by zero-plusplus** [comment]
@thegamerx1
It took some time from the request to the implementation, but I was able to implement it in `1.10.0`.

Thanks for the great idea.
